pkg/core/native_oracle_test.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/nspcc-dev/neo-go/pkg/core/native"
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neo-go/pkg/internal/testchain"
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract/manifest"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract/trigger"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
+	"github.com/nspcc-dev/neo-go/pkg/vm/emit"
+	"github.com/nspcc-dev/neo-go/pkg/vm/opcode"
+	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
+	"github.com/stretchr/testify/require"
+)
+
+// getTestContractState returns test contract which uses oracles.
+func getOracleContractState(h util.Uint160) *state.Contract {
+	w := io.NewBufBinWriter()
+	emit.Int(w.BinWriter, 5)
+	emit.Opcode(w.BinWriter, opcode.PACK)
+	emit.String(w.BinWriter, "request")
+	emit.Bytes(w.BinWriter, h.BytesBE())
+	emit.Syscall(w.BinWriter, interopnames.SystemContractCall)
+	emit.Opcode(w.BinWriter, opcode.RET)
+
+	offset := w.Len()
+	emit.Int(w.BinWriter, 4) // url, userData, code, result
+	emit.Opcode(w.BinWriter, opcode.PACK)
+	emit.Syscall(w.BinWriter, interopnames.SystemBinarySerialize)
+	emit.String(w.BinWriter, "lastOracleResponse")
+	emit.Syscall(w.BinWriter, interopnames.SystemStorageGetContext)
+	emit.Syscall(w.BinWriter, interopnames.SystemStoragePut)
+	emit.Opcode(w.BinWriter, opcode.RET)
+
+	m := manifest.NewManifest(h)
+	m.Features = smartcontract.HasStorage
+	m.ABI.Methods = []manifest.Method{
+		{
+			Name:   "requestURL",
+			Offset: 0,
+			Parameters: []manifest.Parameter{
+				manifest.NewParameter("url", smartcontract.StringType),
+				manifest.NewParameter("filter", smartcontract.StringType),
+				manifest.NewParameter("callback", smartcontract.StringType),
+				manifest.NewParameter("userData", smartcontract.AnyType),
+				manifest.NewParameter("gasForResponse", smartcontract.IntegerType),
+			},
+			ReturnType: smartcontract.VoidType,
+		},
+		{
+			Name:   "handle",
+			Offset: offset,
+			Parameters: []manifest.Parameter{
+				manifest.NewParameter("url", smartcontract.StringType),
+				manifest.NewParameter("userData", smartcontract.AnyType),
+				manifest.NewParameter("code", smartcontract.IntegerType),
+				manifest.NewParameter("result", smartcontract.ByteArrayType),
+			},
+			ReturnType: smartcontract.VoidType,
+		},
+	}
+
+	perm := manifest.NewPermission(manifest.PermissionHash, h)
+	perm.Methods.Add("request")
+	m.Permissions = append(m.Permissions, *perm)
+
+	return &state.Contract{
+		Script:   w.Bytes(),
+		Manifest: *m,
+		ID:       42,
+	}
+}
+
+func TestOracle_Request(t *testing.T) {
+	bc := newTestChain(t)
+	defer bc.Close()
+
+	orc := bc.contracts.Oracle
+	cs := getOracleContractState(orc.Hash)
+	ic := bc.newInteropContext(trigger.System, bc.dao, nil, nil)
```

Comment:
`trigger.Application`? You can put contract's state directly into the `bc.dao`, but regular txes are not run with `trigger.System`.

Reply:
Fixed.